File: tests/test-output.py
import sys
import os
import re
import subprocess
from glob import glob

test_dir = "tests"
luajit_exec = "luajit"
env = os.environ.copy()
env['LUA_PATH']= "./?.lua;;"

if not os.path.isdir("tests/log"):
    try:
        print "Creating directory tests/log..."
        os.mkdir("tests/log")
    except:
        print "Error creating directory tests/log."
        sys.exit(1)

for filename in glob("tests/log/*"):
    os.remove(filename)

try:
	subprocess.check_call([luajit_exec, "-e", ""])
except:
	print "Error calling luajit."
	print "Please make sure that luajit executable is in the current PATH."
	sys.exit(1)

def get_reference_text(dirpath, testname):
    with open(os.path.join(dirpath, testname + ".expect"), "r") as f:
        return f.read()

for dirpath, dirnames, filenames in os.walk(test_dir):
    for filename in sorted(filenames):
        m = re.match(r'([^.]+)\.lua$', filename)
        if m:
            fullname = os.path.join(dirpath, filename)
            testname = m.group(1)
            out_tst = None
            run_error = None
            try:
                process = subprocess.Popen([luajit_exec, fullname], stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env)
                out_tst, out_errors = process.communicate()
            except subprocess.CalledProcessError:
                run_error = "fail to run"
            try:
                out_ref = get_reference_text(dirpath, testname)
            except IOError:
                run_error = "missing expect file"
            led, msg = None, None
            if run_error:
                led, msg = "*", run_error
            elif out_tst == out_ref:
                if out_tst in ["", "\n", "\r\n"] or not out_tst:
                    led, msg = "-", "pass / no output"
                else:
                    led, msg = " ", "pass"
            else:
                led, msg = "*", "fail"
                log = open("tests/log/%s.output.diff" % testname, "w")
                log.write("*** reference ***\n%s\n" % out_ref)
                log.write("*** test program ***\n%s\n" % out_tst)
                log.close()

            print("%s %-24s %s" % (led, testname, msg))

File: test-output.lua
#!/usr/bin/env luajit
local lfs = require("lfs")
local test_dir = "tests"
local log_dir = test_dir.."/log"
package.path = "./?.lua;"..package.path

local function reference_output(dirpath,testname)
  local f, err = io.open(dirpath.."/"..testname..".expect")
  if err then return nil, err end
  local ref = f:read("*a")
  f:close()
  return ref
end

local log_attr = lfs.attributes(log_dir)
if not (log_attr and log_attr.mode=="directory") then
  local ok, err = lfs.mkdir(log_dir)
  if not ok then
    print("Error creating directory "..log_dir..": ",err)
    os.exit(1)
  end
end

local record
do
  local old_print, old_write = print, io.write
  local out_buffer
  local function rec_print(...)
    local args = {...}
    for i,v in ipairs(args) do
      args[i]=tostring(v)
    end
    out_buffer = (out_buffer or "")..table.concat(args,"\t").."\n"
  end
  local function rec_write(...)
    out_buffer = (out_buffer or "")..table.concat({...})
  end
  record = function(cmd)
    if cmd=="stop" then
      print, io.write = old_print, old_write
      return out_buffer
    else
      out_buffer = "" 
      print, io.write = rec_print, rec_write
    end
  end
end

local status = 0
for testfile in lfs.dir(test_dir) do
  local fullname = test_dir.."/"..testfile
  if lfs.attributes(fullname).mode=="file" and testfile:find("%.lua$") then
    local err
    -- Try to load the test file
    local loaded_test, load_err = loadfile(fullname)
    if load_err then
      err = "failed to load: "..load_err
    end
    -- Try to read the reference output 
    local testname = testfile:match("(%S+)%.lua$")
    local out_ref, ref_err = reference_output(test_dir,testname)
    if ref_err then
      err = "missing expect file: "..ref_err
    end
    -- Try to run the test file
    local out_log
    if not err then
      record()
      local run_ok, run_err = pcall(loaded_test)
      out_log = record("stop")
      if not run_ok then
        err = "failed to run: "..run_err
      end
    end
    local led, msg
    if not err and out_log == out_ref then
      if out_log:find("^\r?\n?$") then
        led, msg = "-", "pass / no output"
      else
        led, msg = " ", "pass"
      end
    else
      led, msg = "*","fail"
      status = 1
      if not err then
        local log = io.open(log_dir.."/"..testname..".output.diff", "w")
        if not log then err = "Could not open log file for writing."
        else
          log:write("*** reference ***\n",out_ref,"\n")
          log:write("*** test program ***\n",out_log,"\n")
          log:close()
        end
      end
    end
    print(string.format("%s %-24s %-6s %s",led, testname, msg, err or ""))
  end
end
os.exit(status)
